### ISSUE TITLE
`CopySpec.eachFile(…)` Action not Applied to Copied Directories

### ISSUE THREAD

**Issue by cspurk** [gradle/gradle#2940]
### Expected Behavior

The action given to [`CopySpec.eachFile(…)`](https://docs.gradle.org/current/javadoc/org/gradle/api/file/CopySpec.html#eachFile(groovy.lang.Closure)) should also be applied to each copied directory – at least when [`includeEmptyDirs == true`](https://docs.gradle.org/current/javadoc/org/gradle/api/file/CopySpec.html#eachFile(groovy.lang.Closure)).

### Current Behavior

The action given to `CopySpec.eachFile(…)` is only applied to normal files, even if `includeEmptyDirs == true`.

### Context

In shell terms, I would like to do something as simple as `cp -r source/*/* target/` (i.e., copy all contents of the subdirectories of `source` to `target` – including possibly empty directories). Unfortunately, I couldn’t find any comparably simple way to achieve this in a Gradle `Copy` task, so I tried to remove the leading subdirectory path segment from each copied file in a `CopySpec.eachFile(…)` action.

### Steps to Reproduce (for bugs)

Try the following bash script to quickly set up and execute a simple sample build:
````bash
mkdir test
cd test
mkdir -p source/empty_dir source/path/to/other_empty_dir source/some_dir
touch source/some_dir/file.txt
gradle wrapper --gradle-version 4.1
cat > build.gradle <<'EOF'
task foo(type: Copy) {
    from 'source'
    into 'target'
    includeEmptyDirs = true
    eachFile { println "SEEN: $it" }
}
EOF
./gradlew foo
````

It prints the following:
````
> Task :foo
SEEN: file '/tmp/test/source/some_dir/file.txt'


BUILD SUCCESSFUL in 0s
1 actionable task: 1 executed
````
I would expect to also get “SEEN” lines for at least the empty directories. They are definitely copied:
````
$ tree target/
target/
├── empty_dir
├── path
│   └── to
│       └── other_empty_dir
└── some_dir
    └── file.txt

5 directories, 1 file
````

### Your Environment

````
$ ./gradlew -version

------------------------------------------------------------
Gradle 4.1
------------------------------------------------------------

Build time:   2017-08-07 14:38:48 UTC
Revision:     941559e020f6c357ebb08d5c67acdb858a3defc2

Groovy:       2.4.11
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_131 (Oracle Corporation 25.131-b11)
OS:           Linux 4.4.0-93-generic amd64
````

**Comment by wszeboreq** [comment]
Maybe, in order to avoid the backward compatibility problems, it would be possible to provide an 'eachFile' method variant with additional parameter like e.g. 'includeDirs', so one may use:

```
eachFile(includeDirs: true) {
    println "SEEN: $it"
}
```

We are very interested in having 'eachFile()' variant which works also for directories (lack of it makes it not possible, or at least very hard, to use complete workarounds for the https://github.com/gradle/gradle/issues/1108 problem).

**Comment by lacasseio** [comment]
As part of solving this issue, the caution section user guide chapter "Working with files" should be updated.

**Comment by cspurk** [comment]
Here’s another use case where the feature requested in this ticket would be required:

1. As a workaround for #3982 we have tried to create a custom extension to the `Copy` task which preserves symlinks in the target copy.
2. While this worked for symlinks pointing to regular files, we have run into the issue described in this ticket when copying symlinks pointing to directories. The concrete issue is that we can [`exclude`](https://docs.gradle.org/current/javadoc/org/gradle/api/file/FileCopyDetails.html#exclude--) copied files but we can’t exclude copied _directories_. Because of that, Gradle attempts to set permissions on target subdirectories which are actually source directories below a symlinked directory.
3. Example: when copying `/foo/bar/symlink/subdir/baz.txt` with `symlink` being a symlink to some directory containing `subdir/baz.txt`, then we can exclude `baz.txt` from being copied in the `eachFile` closure but we can’t exclude `subdir/` and hence Gradle tries to set its permissions on the target side that is also the source side in this case. This fails for us because `subdir/` is owned by a different user than the user running Gradle.

**Comment by netvl** [comment]
This should not be closed, because this issue precludes clean extraction of a subtree of an archive.

**Comment by wszeboreq** [comment]
> This should not be closed, because this issue precludes clean extraction of a subtree of an archive.

Exactly, I confirm - see https://github.com/gradle/gradle/issues/1108 and @lacasseio Daniel Lacasse comment https://github.com/gradle/gradle/issues/1108#issuecomment-462977004.

**Comment by netvl** [comment]
BTW, this ticket is even linked to from documentation: https://docs.gradle.org/current/userguide/working_with_files.html#sec:unpacking_archives_example
It seems to me this won't make a good impression on documentation readers.

**Comment by netvl** [comment]
Not planned? Even though it is linked from documentation?

**Comment by ov7a** [comment]
Sorry for the late reply. 

Indeed, this issue should be open, it's still relevant.
